tests/test_directives.py

```diff
@@ -0,0 +1,261 @@
+             'walltime': 64., 'memory': 32, 'processor_fraction': 0.5},
+            {'np': 4, 'ngpu': 1, 'nranks': 0,
+             'omp_num_threads': 10, 'executable': 'PathFinder',
+             'walltime': 20., 'memory': 16, 'processor_fraction': 0.5}]
+
+
+class TestItems:
+    """
+    Tests for _DirectivesItem class
+    """
+    def test_default(self):
+        assert NP.default == 1
+        assert NGPU.default == 0
+        assert NRANKS.default == 0
+        assert OMP_NUM_THREADS.default == 0
+        assert EXECUTABLE.default == sys.executable
+        assert WALLTIME.default == 12.0
+        assert MEMORY.default == 4
+        assert PROCESS_FRACTION.default == 1.
+
+    def test_invalid_values(self, available_directives_list):
+        invalid_values = {'np': [-1, 'foo', {}, None],
+                          'ngpu': [-1, 'foo', {}, None],
+                          'nranks': [-1, 'foo', {}, None],
+                          'omp_num_threads': [-1, 'foo', {}, None],
+                          'walltime': ['foo', {}, None],
+                          'memory': [-1, 'foo', {}, None],
+                          'processor_fraction': [-0.5, 2.5, 'foo', {}, None]
+                          }
+
+        for directive in available_directives_list:
+            if directive.name == 'executable':
+                continue
+            for i, value in enumerate(invalid_values[directive.name]):
+                with pytest.raises((ValueError, TypeError)):
+                    directive.validation(value)
+
+    def test_defaults_are_valid(self, available_directives_list):
+        for directive in available_directives_list:
+            directive.validation(directive.default)
+
+    def test_serial(self):
+        assert NP.serial(4, 2) == 4
+        assert NRANKS.serial(4, 2) == 4
+        assert NGPU.serial(4, 2) == 4
+        assert OMP_NUM_THREADS.serial(4, 2) == 4
+        assert EXECUTABLE.serial('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.serial(4, 2) == 6
+        assert MEMORY.serial(4, 2) == 4
+        assert PROCESS_FRACTION.serial(0.4, 0.2) == 0.4
+
+    def test_parallel(self):
+        assert NP.parallel(4, 2) == 6
+        assert NRANKS.parallel(4, 2) == 6
+        assert NGPU.parallel(4, 2) == 6
+        assert OMP_NUM_THREADS.parallel(4, 2) == 6
+        assert EXECUTABLE.parallel('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.parallel(4, 2) == 4
+        assert MEMORY.parallel(4, 2) == 6
+        assert PROCESS_FRACTION.parallel(0.4, 0.2) == 0.4
+
+    def test_finalize(self):
+        dict_directives = {'nranks': NRANKS.default, 'omp_num_threads': OMP_NUM_THREADS.default}
+        assert NP.finalize(2, dict_directives) == 2
+        dict_directives['nranks'] = 2
+        dict_directives['omp_num_threads'] = 4
+        assert NP.finalize(2, dict_directives) == 2
+        assert NP.finalize(1, dict_directives) == 8
+        dict_directives['nranks'] = lambda x: x**2
```

Comment:
I don't see what this is testing. What's being squared here? (Maybe this needs additional comments in the tests.)

Reply:
This is testing that the `finalize` attribute of `NP` works as expected. This is the method that allows looking at `nranks` and `omp_num_threads` for calculating `np`.

Reply:
The `lambda` function here which is taking x as an argument will take a `job` instead. This function here is just an arbitrary function.

Reply:
I don't think my question was answered. I ran `test_finalize` locally with this line commented out and it still passed. What is this line trying to do? Does it have any effect on the execution? Does this test actually depend on that lambda function being assigned?

Reply:
Technically this was testing if anyone among `nranks` or `opm_num_threads` is callable then return the np passed. But now I think that this test is not necessary. I'll add a more logical test.